DOC-756: Document series column feature for the advtable plugin (#1870)
tinymce/tinymce-docs: _includes/configuration/advtable-value-series.md

## `advtable_value_series`

> {{ site.requires_5_8 }}

Reviewer: Should this be 5.7? As the code currently will allow it for 5.7 or higher.

{% if pluginname != "Advanced Tables" %}
> **Note**: The `advtable_value_series` option requires the Advanced Tables plugin.
{% endif %}

The `advtable_value_series` option allows values series to be configured that can be used to populate cells in a table.

**Type:** `Object`

**Default Value:**
 - `numbers` - a natural number series.

>  - `alpha` - a base26 letter series.

Reviewer 2: I feel like adding the info that it's base26 makes it more confusing. Could we just call it "an alphabetic series" or something like that (maybe with a note that it uses the English alphabet)?

```js
{
  numbers: {
    update: true,
    resizable: false,

>     generator: ...

Reviewer 3: will this take a function?

Reviewer 2: Do we need to write the function out here, or is it a common pattern in the docs to put `...` into code snippets?

Reviewer: According to the [style guide](https://www.tiny.cloud/docs/configure/contributing-docs/#codesnippets), `...` in code samples is bad and shouldn't be used.

Reviewer 4: yeah, please provide a full example. Ideally, users should be able to copy+paste codeblocks into, say, fiddle and have the code work without modification (excluding having to change the selector).

Author: Have removed the `...`

  },
  alpha: {
    update: true,
    resizable: false,
    generator: ...
  },
}
```

### Series configuration

| Name | Value | Requirement | Description |
| ---- | ----- | ----------- | ----------- |
| update | `boolean` | Optional | default: `false` - Specifies whether the table cells that are part of the series should update after changes are made to the table. |
| resizable | `boolean` | Optional | default: `true` - Specifies whether the table cells that are part of the series are resizable when using the resize bars. |
| generator | `(info: GeneratorInfo, rowIndex: number, columnIndex: number) => GeneratorResult` | Required | [Usage of generator](#usageofgenerator). |

#### Usage of generator

The `generator` is a callback function that is used to specify how a table cell that is part of the series should be updated. The callback is passed information relating to the generator and table cell, the row index and column index of the table cell. For more details, see: [GeneratorInfo](#generatorinfo).

> The callback should return an object containing the value and optionally, the classes and attributes that should be applied on the table cell. For more details, see: [GeneratorResult](#generatorresult).

Reviewer 2: I think this either should be joined onto the previous line (if it's part of the same paragraph) or it should have an extra newline before it (to make it a separate paragraph).

##### GeneratorInfo

| Name | Value | Description |
| ---- | ----- | ----------- |

> | rowType | `'thead'`, `'tbody'` or '`tfoot'` | The section of the table cell. |

Reviewer 2: ```suggestion
| rowType | `'thead'`, `'tbody'` or `'tfoot'` | The section of the table cell. |
```

Reviewer 2: Also, do we normally use the Oxford comma in these situations? From notion, it looks like marketing uses it, but I don't know about docs.

| cellType | `'td'` or `'th'` | The type of the table cell. |
| direction | `'row'` or `'column'` | The direction of the generator. |
| prevSeriesValue | `string` or `undefined` | The previous value calculated from the generator. |

##### GeneratorResult

| Name | Value | Requirement | Description |
| ---- | ----- | ----------- | ----------- |
| classes | `string[]` | Optional | The classes that should be applied on the table cell. |

> | attributes | `Record<string, string, boolean, number or null>` | Optional | The attributes that should be applied or removed on the table cell. |

Reviewer 2: The `Record<...>` section needs to be reworked a bit. There's some ambiguity around what the `,` means because it could be `Record<type one, type two>` but it could also mean `type one, type two or type three`? The only potential solution I can think of is to split it out more, like this:

```
Record<string, string>, Record<string, boolean>, Record<string, number> or Record<string, null>
```

But that (1) is super wordy and (2) implies that you can't mix the values up within the same object.

Maybe we could just say that the type is `Object` and then in plain text on what values can be put inside the object?

Reviewer 2: Also, I'm guessing that `{ 'my-attribute': null }` will remove `my-attribute` from the cell, but that should probably be stated a little more explicitly if possible.

Reviewer: Worth noting I don't believe we normally say `Record` in the docs and instead use `Object`.

Author: I have changed the `value` column to be `Object` and updated the `description` column accordingly.

| value | `string`, `number` or `undefined` | Optional | The value of the table cell. If the given value is `undefined`, it will attempt to use the previous value of the table cell. |

### Example: Using `advtable_value_series`

```js
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: 'table advtable',
  menubar: 'table',
  advtable_value_series: {
    numbers: {
      update: true,
      resizable: false,
      generator: function (info, rowIndex, columnIndex) {
        return {
          value: rowIndex + 1
        };
      }
    },
  }
});
```

> {% comment %} TODO: Do I need to make a note that this option is used in conjuction with the mceTableToggleSeries command? {% endcomment %}

Reviewer 3: if they don't define advtable_value_series, and call mceTableToggleSeries, does it throw some kind of advtable_value_series is undefined?

Reading this on the one pager, its implied that we need that setting to make it work, on the actual website, that may not be so clear.  We could note that in advtable-cmds.md, requires setting advtable_value_series or throw advtable_value_series needs to be defined in the configuration

Reviewer 3: oh the next advtable.md, says just that

> {% comment %} TODO: Do I need to make a note that this option is used in conjuction with the mceTableToggleSeries command? {% endcomment %}

Reviewer 2: I'd say add that as a sentence after "The `advtable_value_series` option allows values series to be configured that can be used to populate cells in a table."

## `advtable_value_series`
{{ site.requires_5_7 }}

{% if pluginname != "Advanced Tables" %}
> **Note**: The `advtable_value_series` option requires the Advanced Tables plugin.
{% endif %}

The `advtable_value_series` option is used in conjuction with the [mceTableToggleSeries]({{site.baseurl}}/plugins/premium/advtable/#commands) command.

> The `advtable_value_series` option allows values series to be configured that can be used to populate cells in a table.

Reviewer 3: The `advtable_value_series` option configures the incremental values in a series that will be used to populate cells in a table

**Type:** `Object`

**Default Value:**

```js
{
  // Natural number series
  numbers: { 
    update: true,
    resizable: false,

>     generator: `GeneratorFunction` // For deatils, see: 'Usage of generator'

Reviewer: ```suggestion
    generator: `GeneratorFunction` // For details, see: 'Usage of generator'
```
Minor typo here

>     generator: `GeneratorFunction` // For details, see: 'Usage of generator'

Reviewer 4: ```suggestion
    generator: `GeneratorFunction` // For details, see: 'Creating a value series generator'
```

Author: Oops. Thanks for picking this up.

  },
  // English alphabetic series
  alpha: {
    update: true,
    resizable: false,

>     generator: `GeneratorFunction` // For deatils, see: 'Usage of generator'

Reviewer: ```suggestion
    generator: `GeneratorFunction` // For details, see: 'Usage of generator'
```

>     generator: `GeneratorFunction` // For details, see: 'Usage of generator'

Reviewer 4: ```suggestion
    generator: `GeneratorFunction` // For details, see: 'Creating a value series generator'
```

  },
}
```

Both default series are configured to update on table changes and not resize when using the resize bars.

### Series configuration

| Name | Value | Requirement | Description |
| ---- | ----- | ----------- | ----------- |

> | update | `boolean` | Optional | default: `false` - Specifies whether the table cells that are part of the series should update after changes are made to the table. |

Reviewer 4: ```suggestion
| update | `boolean` | Optional | default: `false` - When `true`, the series values will be updated when changes are made to the table. |
```

> | resizable | `boolean` | Optional | default: `true` - Specifies whether the table cells that are part of the series are resizable when using the resize bars. |

Reviewer 4: ```suggestion
| resizable | `boolean` | Optional | default: `true` - When `true`, table cells containing the series values can be resized using a mouse or touch device. |
```

> | generator | `(info: GeneratorInfo, rowIndex: number, columnIndex: number) => GeneratorResult` | Required | [Usage of generator](#usageofgenerator). |

Reviewer 4: ```suggestion
| generator | `(info: GeneratorInfo, rowIndex: number, columnIndex: number) => GeneratorResult` | Required | For details on creating a value series generator, see: [Usage of generator](#usageofgenerator). |
```
May also need to update link

#### Usage of generator

> The `generator` is a callback function that is used to specify how a table cell that is part of the series should be updated. The callback is passed information relating to the generator and table cell, the row index and column index of the table cell. For more details, see: [GeneratorInfo](#generatorinfo). The callback should return an object containing the value and optionally, the classes and attributes that should be applied on the table cell. For more details, see: [GeneratorResult](#generatorresult).

Reviewer 4: ```suggestion
The `generator` is a callback function for specifying how to update a table cell of a value series. The callback is passed information relating to the generator and table cell, the row index, and column index of the table cell. For details, see: [GeneratorInfo](#generatorinfo). The callback should return an object containing the value and optionally, any classes and attributes to be applied to the table cell. For details, see: [GeneratorResult](#generatorresult).
```

##### GeneratorInfo

| Name | Value | Description |
| ---- | ----- | ----------- |
| rowType | `'thead'`, `'tbody'` or `'tfoot'` | The section of the table cell. |
| cellType | `'td'` or `'th'` | The type of the table cell. |
| direction | `'row'` or `'column'` | The direction of the generator. |

> | prevSeriesValue | `string` or `undefined` | The previous value calculated from the generator. |

Reviewer 4: ```suggestion
| prevSeriesValue | `string` or `undefined` | The previous value calculated by the generator. |
```

##### GeneratorResult

| Name | Value | Requirement | Description |
| ---- | ----- | ----------- | ----------- |

> | classes | `string[]` | Optional | The classes that should be applied on the table cell. |

Reviewer 4: ```suggestion
| classes | `string[]` | Optional | The classes to be applied to the table cell. |
```

> | attributes | `Object` | Optional | The attributes that should be applied on the table cell. `attributes` should be specified as an object where each key is an attribute and each value is of type `string`, `boolean`, `number` or `null`. If `null` is given as the value for an attribute, the attribute is removed from the table cell. |

Reviewer 4: ```suggestion
| attributes | `Object` | Optional | The attributes to be applied to the table cell. The `attributes` should be provided as an object where each key is an attribute and each value is of type `string`, `boolean`, `number` or `null`. A value of `null` for an attribute will remove the attribute from the table cell. |
```

> | value | `string`, `number` or `undefined` | Optional | The value of the table cell. If the given value is `undefined`, it will attempt to use the previous value of the table cell. |

Reviewer 4: ```suggestion
| value | `string`, `number` or `undefined` | Optional | The value of the table cell. If the value is `undefined`, the editor will use the previous value of the table cell. |
```
